docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md

```diff
@@ -0,0 +1,123 @@
+    * All application-services changes that are breaking for consumers should have corresponding PRs in the consumer repos that have been reviewed and approved.  If we accidentally merge the change into application-services before the PRs are ready, then we should back the change out.
+
+## Decision Outcome
+
+**(B) Move to a rapid release process where changes in app-services are frequently merged into consumer applications**
+
+## Pros and Cons of the Options
+
+### Keep our current release process
+
+  * Good, because it doesn't require changes to our workflow
+  * Bad, for the reasons enumerated at the top of this document
+
+### Move to a rapid release process where changes in app-services are frequently merged into consumer applications
+
+  * Good, because code will get merged and released more frequently.
+  * Good, because we don't need to spend time cutting releases.
+  * Good, because uplifts can be done much quicker.
+  * Good, because more of our release process will be automated.
+  * Bad, because we need implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
+  * Good, because this avoids the current state of affairs where our `main` branch is not ready to be taken by the applications.  This means that a) our current code is not getting used and tested and b) it's very complicated to get other changes to that application, including bugfixes.
+  * Bad, because breaking changes require the corresponding PRs to be approved from all other teams before they can be merged into application-services.
+  * Good, because this is how things will work in a monorepo world.  Adjusting our workflow now prepares us for the future and will give us insights that may help guide the monorepo migration.
+  * Good, because our version numbers will align more closely with the rest of Mozilla.
+  * Good, because it makes it easier to know which application-services version the applications are using.  The current Firefox X release will always be using the latest application-services.X release -- for both the mobile apps and Desktop.
+
+# Non-normative discussion
+
+## Impact on other teams
+### Application teams -- Android, iOS, and Desktop
+
+* Phase 1
+  * The application-services version number style will change
+  * There may be a increase in the amount of application-service version upgrade PRs to approve.  We'll want to make sure that each application is on an application-services release that matches their release (i.e. avoiding having Fenix 110.x on application-services 109.x).  For the Android/iOS teams this will only be a small increase, but for Desktop it will be larger.
+* Phase 2
+  * More application-services PRs to approve, especially at the start.  Eventually, we should be able to arrange it so most PRs get auto-merged.
+
+### Release Management
```

Comment:
I've very hesitant to sign off on our team taking this on without having a better understanding of what the workflow would be here. It sounds like these steps are intended to be mostly automated per the plan above? Is this just going to be a matter of triggering a hook somewhere, starting a build in ship-it, or something else?

Reply:
And from an expectations standpoint, for releases, is the expectation that a new point release will be created for every uplift to a release branch or done on some schedule? Or when requested by the team?

Reply:
> I've very hesitant to sign off on our team taking this on without having a better understanding of what the workflow would be here. It sounds like these steps are intended to be mostly automated per the plan above? Is this just going to be a matter of triggering a hook somewhere, starting a build in ship-it, or something else?

I don't think we know exactly how this is going to work, and I wanted to leave those details out of the ADR, but  it should certainly be automated.  The work should be something like triggering a hook and monitoring that it succeeds.  I wasn't sure if that should be our team's responsibility or relman, I'm happy to follow your recommendation here.

> And from an expectations standpoint, for releases, is the expectation that a new point release will be created for every uplift to a release branch or done on some schedule? Or when requested by the team?

With our current workflow, we backport fixes about once a month or less.  Based on that, I think we should make a new point release as soon as the uplift is merged since it's unlikely that any more uplifts are coming soon.

Reply:
We can take this discussion outside the ADR if you want. But my high-level recommendation would be to utilize Ship-It for creating releases and have those be self-serviced by the A-S team. RelMan would then be involved as usual with approving uplifts to the appropriate release branches after the new release is ready. We can talk through the details in greater depth later, but this has some strong advantages over manually tagging releases in GitHub or triggering hooks in TC to kick off a new release and aligns well with the overall direction we're trying to move in around consolidating releases through there.